Implementation multi pattern regular expressions

biathlon3 opened the pull request:
To compile regexp **required** installation of hscollider. For simplicity it built as debian package.

Add repository
```
sudo install -m 0755 -d /etc/apt/keyrings
sudo wget -O /etc/apt/keyrings/tempesta.asc https://tempesta-tech.com:8081/repository/public-keys/keys/public.key
sudo chmod a+r /etc/apt/keyrings/tempesta.asc
echo "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/tempesta.asc] \
https://tempesta-tech.com:8081/repository/tempesta/ focal main" | sudo tee /etc/apt/sources.list.d/tempesta.list
sudo apt update && sudo apt upgrade
```

Install the package
```
apt install linux-regex
```

Commit:
Implemented regex for locations and httptables

For regular-expression matching, use ~ for locations and the HTTP chain.

Example:
```
location ~ "/^/show/id/[0-9]+/i" {
    frang_limits {
	http_body_len 5;
	http_strict_host_checking true;
    }
}

http_chain {
  uri ~  "^/html/" -> default;
  -> default;
}
```

biathlon3 commented:
For deployment see file install.txt. Description will be expanded.

biathlon3 commented:
This PR works on Kernel 6.8.9

Description of using hscollider - http://intel.github.io/hyperscan/dev-reference/tools.html

Description of regular expression syntax - https://perldoc.perl.org/perlre
It is supported in Hyperscan not fully, constraints described here - http://intel.github.io/hyperscan/dev-reference/compilation.html

Now regex implemented in tempasta config as fallows: 
for locations, added keyword "regex" and regex started with "^";
for httptables, regex started with "^".

for example:
```
location regex "^/new/" {
    frang_limits {
	http_body_len 5;
	http_strict_host_checking true;
    }
}

http_chain {
  uri == "^/html/" -> default;
  -> default;
}
```

biathlon3 commented:
I squashed it into one commit so it can be reviewed.

const-t committed:
fix: make kfuns available

const-t committed:
fix: nested kernel_fpu_begin()

`rex_scan_tfwstr()` is intended to be called in softIRQ
context, therefore don't call `kernel_fpu_begin()`
because it being called before handling softIRQ.

const-t committed:
fix: alignment

const-t committed:
fix: remove regex setup logic from tempesta.sh

Also change path to regex folder.

const-t committed:
fix: allow apply regular expressions to raw headers using HTTPTables

Beside the part which allows to apply regular expressions to raw
headers this patch changes string matching for raw headers. Matching
consist of two parts: matching a header name and matching a header
value. Name always matches by simple string comparison, when this
part is successfull Tempesta matches value by string comparasion or
using regular expressions.

Name matching relies on header validity that guaranteed by parser,
it always assume that name of a header doesn't have OWS for HTTP1 and
always lower case for HTTP2. OWS are allowed only for value and only
for HTTP1. For HTTP2 we use memcmp_fast() to compare chunks of header
name, because a lot of small chunks not expected for HTTP2, we place
headers into big contigious chunks during hpack decoding. However, for
HTTP1 we use __str_cmp() that is simplified version of strncasecmp(),
but doesn't do tolower() for pattern, only for headwer part. Pattern
for string comparison now always lower case and space trimmed during
configuration stage on startup of Tempesta.

Also I did my best to get rid of branches inside the loops where it
is possible, making function inline with conditional argument known at
compile time. Alternatives are indirect calls and macroses: the former
might be slow (they are in my benchmarks) the second make the code
hard to read.

const-t committed:
feat: remove spaces from header pattern

const-t committed:
chor: Remove unused enum field and structure field

const-t committed:
tests: add couple of tests and fix already existing

const-t committed:
refactor: remove unused fields

const-t committed:
refactor: use kernel's min/max

const-t committed:
fix: enable avx512

Enabling avx512 causing compilation time wanings relateed to
string fortification. Compiler can't check the size of the data,
because the length of the data calculates at runtime. Do unsafe
copy to suppress the warnings.

const-t committed:
feat: compile regex with the same standard as kernel

const-t committed:
fix: remove *~ token, instead use /i modifier in the pattern

const-t committed:
fix: remove unused argument and add validation

const-t committed:
chore: remove unused file

const-t committed:
fix: replace kernel allocation functions by Tempesta wrappers

const-t committed:
fix: change style

const-t committed:
fix: add error handling when writing regexp file

const-t commented:
Description of using hscollider - http://intel.github.io/hyperscan/dev-reference/tools.html

Description of regular expression syntax - https://perldoc.perl.org/perlre
It is supported in Hyperscan not fully, constraints described here - http://intel.github.io/hyperscan/dev-reference/compilation.html

Use `~` for regexp in location and http chain:
```
location ~ "/^/show/id/[0-9]+/i" {
    frang_limits {
	http_body_len 5;
	http_strict_host_checking true;
    }
}

http_chain {
  uri ~  "^/html/" -> default;
  -> default;
}
```

const-t committed:
feat: add config rules to block clients with suspicious user-agents

The config can be included using !include directive in http_chain
section. Example:

```
http_chain {
    #some rules ...
    !include /etc/tempesta/ua_block_rules.conf
}
```

const-t committed:
fix: match strictly only "!include path"

Do strict matching to not match not suitable strings like "!includeasd"
and "   #include"

const-t committed:
feat: add user-space helper for regexp compilation

What has been done:
 - Added a user-space helper `regex_setup.sh` that passes expressions
   to hscollider and copies the compiled database to the rex module
   database. Path to the script can be specified using environment
   variable REGEX_SETUP_SCRIPT_PATH and path to the regex databases
   can be specified using REGEX_DIR_PATH variable. Implemented as
   kernel module parameter.
 - Moved regular expression code to its own file.
 - Rex directory added to `rules` file.

const-t committed:
fix: Allocate only required amount of memory for the regex ID

In this patch we allocate only the amount of memory needed for the
regex ID, instead of allocating memory for the full string pattern

EvgeniiMekhanik commented:
I catch memory leak in case when configuraion fails

kmemleak_alloc+0x4a/0x90
    __kmalloc_noprof+0x3a5/0x520
    tfw_kzalloc+0x14/0x20 [tempesta_lib]
    rex_make_item+0x27/0x100 [xdp_rex]
    configfs_mkdir+0x735/0xd30
    vfs_mkdir+0x45c/0x7a0
    do_mkdirat+0x292/0x320
    __x64_sys_mkdir+0xd7/0x130
    x64_sys_call+0x10cf/0x25f0
    do_syscall_64+0x7e/0x170
    entry_SYSCALL_64_after_hwframe+0x76/0x7e

config:

listen 80;
listen 443 proto=h2;
        
tls_certificate /home/evgeny/workdir/certs/tempesta-tech.test.crt;
tls_certificate_key /home/evgeny/workdir/certs/tempesta-tech.test.key;
tls_match_any_server_name;

access_log off;

srv_group grp1 {
    server 127.0.0.1:8000;
}

vhost test {
    proxy_pass grp1;
    location ~ "/^/article/id/[0-9]+/" {
        proxy_pass grp1;
    }
    location ~ "/shop/" {
        proxy_pass grp1;
    }
    location ~ "/wiki/" {
        proxy_pass grp1;
    }
}

vhost work {
    proxy_pass grp1;
    location ~ "/shop/i" {
        proxy_pass grp1;
    }
    location ~ "/wiki/" {
        proxy_pass grp1;
    }
}

http_chain {
    host ~ "/test/i" -> test;
    host ~ "/work/i" -> work;
    host ~ "/(a+)+$$/" -> work;
    -> block;
}

EvgeniiMekhanik commented on fw/http_match.c:
```diff
@@ -827,6 +827,11 @@ tfw_http_arg_adjust(const char *arg, tfw_http_match_fld_t field,
 	if (wc_arg && !raw_hdr_name)
 		return NULL;
 
+	if (wc_arg && regex) {
```
host ~ "*" -> test -> block still work here, because raw_hdr_name is NULL, is it ok?

EvgeniiMekhanik commented on scripts/regex_setup.sh:
```diff
@@ -0,0 +1,76 @@
+# this program; if not, write to the Free Software Foundation, Inc., 59
+# Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+
+regex_dir_path=${REGEX_DIR_PATH:="/opt/tempesta/regex"}
```
It seems that you set `regex_dir_path` in regex.c without any chance to change it.

const-t replied:
REGEX_DIR_PATH is env var, tempesta.sh passes it to kernel and then kernel sets is when calls helper

EvgeniiMekhanik commented on fw/http_match.c:
```diff
@@ -835,7 +905,13 @@ tfw_http_arg_adjust(const char *arg, tfw_http_match_fld_t field,
 	if (raw_hdr_name && field != TFW_HTTP_MATCH_F_COOKIE) {
 		name_len = strlen(raw_hdr_name);
-		full_name_len = name_len + SLEN(S_DLM);
+		full_name_len = name_len + SLEN(S_COLON);
```
Why `full_name_len` is not initialized for TFW_HTTP_MATCH_F_COOKIE

EvgeniiMekhanik commented:
I also can't understand
if I add
hdr cookie ~ "/fff*/" -> block; in http_chain should request with  ("cookie", "fffoooo=bbb"), blocked?

const-t committed:
fix: alignment

const-t committed:
fix: forbid using wildcard with regexp

const-t committed:
fix: regex length validation

Remove outdated comment and change minimal regex length. Currently
we allow even 1 symbol regex, but not empty regex.

const-t commented on fw/vhost.c:
```diff
@@ -1528,23 +1532,20 @@ tfw_location_init(TfwLocation *loc, tfw_match_t op, const char *arg,
-		unsigned short regex_idx;
 
-		if ((r = tfw_write_regex(arg, &regex_idx))) {
+		r = tfw_write_regex(arg, &regex_idx);
```
Fixed comment and changed condition

const-t committed:
refactor: add comment and change arguments name

const-t committed:
feat: compile avx/avx2/avx512 only if it is supported

const-t committed:
feat: enable avx512vbmi

const-t committed:
fix: memory leak during config_item release

const-t committed:
fix: ensure that cookie match doesn't used with regex

`cookie "cookie_name" ~ "cookie_value"` is not supported at this moment,
seems it doesn't make sense to have such option, matching by header
could be useed instead to find all occurrences in single regex call.
Example:
`hdr cookie ~ "/cookie_name=cookie_value/"`

const-t commented on regex/kmod/rex.c:
```diff
@@ -243,64 +190,46 @@ int bpf_scan_tfwstr(const TfwStr *str,
-
-	if (offset + len > size)
-		return ERR_PTR(-EINVAL);
+EXPORT_SYMBOL(bpf_scan_bytes);
```
Yes, you are right

EvgeniiMekhanik commented on fw/t/unit/regex.c:
```diff
@@ -0,0 +1 @@
+../../regex.c
```
No empty line at the end of the file?

EvgeniiMekhanik reviewed (approved):
Please also add three stress tests with reconfiguration with regexp:
- With success reconfiguration regexp config - regexp config
- Fail reconfig
- Success was regexp now not regexp

krizhanovsky commented on fw/http_match.c:
```diff
@@ -84,6 +85,7 @@ map_op_to_str_eq_flags(tfw_http_match_op_t op)
 		[TFW_HTTP_MATCH_O_EQ]		= TFW_STR_EQ_DEFAULT,
 		[TFW_HTTP_MATCH_O_PREFIX]	= TFW_STR_EQ_PREFIX,
 		[TFW_HTTP_MATCH_O_SUFFIX]	= TFW_STR_EQ_DEFAULT,
+		[TFW_HTTP_MATCH_O_REGEX]	= TFW_STR_EQ_REGEX
```
Please add some info to the HTTP tables wiki, make good examples

const-t replied:
added https://github.com/tempesta-tech/tempesta/wiki/HTTP-tables

krizhanovsky commented on fw/http_tbl.c:
Add `~` to the operation description in the comment

const-t replied:
added

krizhanovsky commented on fw/http_match.c:
```diff
@@ -819,6 +819,11 @@ tfw_http_arg_adjust(const char *arg, tfw_http_match_fld_t field,
 
+	if (wc_arg && regex) {
+		T_ERR_NL("http_match: use simple wildcard argument: hdr == *. Instead regex.\n");
+		return ERR_PTR(-EINVAL);
```
```suggestion
		T_ERR_NL("http_match: use simple wildcard argument `hdr == *`"
		         " instead of regex.\n");
```
By the way

krizhanovsky commented on fw/vhost.c:
```diff
@@ -1421,6 +1421,11 @@ tfw_cfgop_out_cache_control_ignore(TfwCfgSpec *cs, TfwCfgEntry *ce)
+ * NOTE: This works only for non-regex locations. Regex locations may have the
+ * same pattern. It seems not worth trying to detect duplicate patterns,
+ * because two similar but not identical patterns can still resolve to the same
+ * location.
```
Please add regular expressions for `location` to wiki

const-t replied:
I've added. https://github.com/tempesta-tech/tempesta/wiki/Virtual-hosts-and-locations

krizhanovsky commented on regex/kmod/rex.c:
Please add our copyrights to all changed files like regex/kmod/rex.c

krizhanovsky commented on fw/t/unit/test_http_match.c:
```diff
@@ -913,7 +1010,9 @@ TEST_SUITE(http_match)
+	TEST_RUN(http_match, raw_header_eq_nows);
 	TEST_RUN(http_match, method_eq);
 	TEST_RUN(http_match, cookie);
 	TEST_RUN(http_match, choose_host);
```
Could you please add unit tests for regular expressions

krizhanovsky commented on install.txt:
```diff
@@ -0,0 +1,42 @@
+
+git clone https://github.com/tempesta-tech/tempesta.git
+cd tempesta
+git checkout ag_Multi-pattern-regular-expressions
```
I think this should be README.md in the `regex` directory and reference the original https://github.com/G-Core/linux-regex-module to give credits to the team. Probably this wiki https://github.com/tempesta-tech/tempesta/wiki/Installation-tempesta-supported-regex should be in the README - it's useful for us, but not for users (can we remove the wiki page after that?)

krizhanovsky commented on scripts/tempesta.sh:
```diff
@@ -58,6 +61,7 @@ lib_mod=tempesta_lib
 tls_mod=tempesta_tls
 tdb_mod=tempesta_db
 tfw_mod=tempesta_fw
+rgx_mod=xdp_rex
```
That project was for the XDP module and technically could be useful in https://github.com/tempesta-tech/xFWas , but anyway the regexps, especially for Tempesta FW, have nothing to do about XDP. Let's rename the module to something like `tempesta_regex` that an administrator can easily see all our modules

krizhanovsky commented on fw/http_match.c:
```diff
@@ -767,7 +767,7 @@ write_regex(const char *arg, int regex)
 
 	++number_of_db_regex;
-	sprintf(file_name, "/tmp/tempesta/%u.txt", number_of_db_regex);
+	sprintf(file_name, "/opt/tempesta/regex/%u.txt", number_of_db_regex);
```
We should get the path from the script: this path is likely to change, IIRC @symstu recently made a change in Escudo that all the products are installed in `/opt/tempesta` and have their own subdirs, `fw`, `xfw`, `webshield`. We can do this with sysctl or module parameter, but this should be a generic home/work directory specification

const-t replied:
Maybe you commented old version, currently this path is passed to module using parameters.

krizhanovsky commented on regex/kmod/rex.c:
```diff
@@ -91,12 +90,12 @@ static int rex_scan_cb(unsigned int expression, unsigned long long from,
 }
 
-int bpf_scan_bytes(const void *buf, __u32 buf__sz, struct rex_scan_attr *attr)
+int rex_scan_tfwstr(const TfwStr *str, struct rex_scan_attr *attr)
```
Now the code is in our tree, so please adjust coding style of all new and changed functions and definitions

krizhanovsky commented on regex/kmod/rex.c:
```diff
@@ -119,60 +118,11 @@ int bpf_scan_bytes(const void *buf, __u32 buf__sz, struct rex_scan_attr *attr)
-}
-EXPORT_SYMBOL(bpf_scan_bytes);
-
-int bpf_scan_vector(const char *const *buf,
```
I think there is no reason to remove the BPF-purpose routines - they might be useful in xFW, but for sure we don't need code for old kernel versions

const-t replied:
I removed this this function because it has been added by us in this commit https://github.com/tempesta-tech/tempesta/pull/2161/changes/4d5e3d48d0d74d0f7cef00b4c6d9febe96d05865 and this code is not test covered and even not being used, I believe this is bad approach to leave the code which we are not sure about.

krizhanovsky commented on fw/regex.c:
```diff
@@ -0,0 +1,240 @@
+#include "regex/kmod/rex.h"
+#include "tempesta_fw.h"
+
+static char *regex_setup_script_path = "/lib/tempesta/scripts/regex_setup.sh";
```
This path is also installation-dependent, we should get it from the outside: sysctl, module param etc

const-t replied:
Maybe you commented old version, currently this path is passed to module using parameters.

krizhanovsky commented on fw/http_match.c:
```diff
@@ -242,174 +242,243 @@ match_host(const TfwHttpReq *req, const TfwHttpMatchRule *rule)
-		char prev;
-		short cnum;
+	return 0;
+}
```
Isn't it the same as `tfw_cstricmp_2lc()`?

const-t replied:
The main difference is `tfw_cstricmp_2lc()` always calls `tolower()` for both arguments `string` and `pattern` but `__cmp_hdr_raw_name()` doesn't call `tolower()` for h2 at all and do `memcpy_fast()`. For h1 `tolower()` will be called only for `string` not for a substring.

krizhanovsky reviewed (approved):
A few improvements are possible, but looks good to merge in general.

I appreciate `match_hdr_raw()` refactoring!

const-t committed:
refactor: Add regexp description to doc comment

const-t committed:
refactor: prettify error message

const-t committed:
chore: add copyright to modified files

const-t committed:
refactor: change file formatting to Tempesta coding style

const-t committed:
chore: change regex module name

const-t committed:
chore: remove unused files

const-t committed:
chore: Add readme for regex module

const-t committed:
feat: add regexp unit tests

const-t commented:
@krizhanovsky PR is updated. Please consider all review comments addressed.
